### PR TITLE
FEC-12022 Version 0.4.0

### PR DESCRIPTION
**Pull request by GouravSna**
- Solves [FEC-12643](https://kaltura.atlassian.net/browse/FEC-12643)

- **RN Bridge** 
       - Updated `react native version >= 0.66.5` in order to avoid `Hermes-Engine v0.9.0` vulnerability and Android build issue on that version.
       - **[Android]** Sending `PlayerEvent` `Error` in all the cases for `android`; for `iOS` it is already there. FE can determine the severity. 
       - **[iOS && Android]** Fixed `error` event for Ad bug. Error was coming twice (for player event error also it was coming because for RN it is `string` in the listener)
       - **[Android]** Added `DRMSettings` which gives ability to play `Playready` media as well.
       - **[iOS && Android]** Made `mediaAsset` for `loadMedia` optional on bridge and native level.
       - **[iOS && Android]** Made `initOptions` optional on `setup` player on bridge and native level.
       - **[Android]** Added **start position** for OVP/OTT Player.
       - **[Android]** Enable `enableDebugLogs` API on bridge, it on/off the Playkit and Netkit logs. For **iOS**, it is not enabled in this version.
       - **[iOS]** Converting it to millis in `AdEvents.Loaded` because Android is also send it in millies.
       - **[Android]** Added BroadPeak error event on Android. 
       - **[iOS && Android]** Added `BROADPEAK_ERROR` on RN bridge.
       -  **[iOS]** Added No-Op Analytics events on iOS side without implementation to avoid crash due to unavailable events.
       - **[iOS && Android]** Renamed `error` to `phoenixError` to avoid name clash with player events. This will not make any code change on client app side.

- **Example App Fixes**
       - Upgraded `react native version to 0.68.5` on example app. It fixes the issue which happened [here](https://github.com/facebook/react-native/issues/35210).
       - **[Android]** Updated `flipper` version.
       - Progress bar on player screen will be updated with a timer only for DAI enabled streams. 
       - Fixed UI for Ads in case of iOS.
       - Added multi language media and fixed the audio language name in drop down.
       - If label is not there, take language for audio and text track.

### CODE REVIEW

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt [hunk 1]
```diff
@@ -526,52 +530,43 @@ class KalturaPlayerRN(
...
-//        when (logLevel) {
-//            "VERBOSE" -> {
-//                PKLog.setGlobalLevel(PKLog.Level.verbose)
-//                NKLog.setGlobalLevel(NKLog.Level.verbose)
-//                YouboraLog.setDebugLevel(YouboraLog.Level.VERBOSE)
-//            }
-//            "DEBUG" -> {
-//                PKLog.setGlobalLevel(PKLog.Level.debug)
-//                NKLog.setGlobalLevel(NKLog.Level.debug)
-//                YouboraLog.setDebugLevel(YouboraLog.Level.DEBUG)
```
Comment (GouravSna) [inline android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt:547]: Put this in the doc, how to enable youbora.

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt [hunk 2]
```diff
@@ -849,16 +846,15 @@ class KalturaPlayerRN(
...
-    private fun isBasicPlaybackRequired(basicMediaAsset: String?): Boolean {
-        if ((basicMediaAsset != null) && (playerType != null) && (playerType != KalturaPlayer.Type.basic)) {
-            val mediaAsset = getParsedJson(
-                    basicMediaAsset,
-                    BasicMediaAsset::class.java
-            )
-            return mediaAsset != null && mediaAsset.mediaFormat != null
+    private fun isBasicPlaybackRequired(assetURL: String?): Boolean {
+        if (!TextUtils.isEmpty(assetURL) && playerType != null && playerType != KalturaPlayer.Type.basic) {
+            val isBasicPlayback = URLUtil.isValidUrl(assetURL) && PKMediaFormat.valueOfUrl(assetURL) != null
```
Comment (GouravSna) [inline android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt:863]: PKMediaFormat.valueOfUrl(assetURL), What if there is no media extension in URL and there is valid mediaAsset having media format. Need to raise a ticke to fix such scenario.

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt [hunk 3]
```diff
@@ -654,6 +647,7 @@ class KalturaPlayerRN(
                             if (mediaAsset.initialVolume >= 0 && mediaAsset.initialVolume < 1.0) {
                                 player?.setVolume(mediaAsset.initialVolume)
                             }
+                            player?.startPosition = mediaAsset.startPosition
```
Comment (GouravSna) [inline android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt]: Check the DVR live media with Start Position.

File: docs/events.md [hunk 4]
```diff
@@ -152,6 +152,12 @@ It talks about more about the concurrency level errors on the stream. These even
 
 |AnalyticsEvents         |Data sent to RN |  iOS | Android |
 |------------|--------|----------|----------|
-|`PHOENIX_CONCURRENCY_ERROR`|**TODO**|✅ |✅ |
-|`PHOENIX_BOOKMARK_ERROR`|**TODO**|✅ |✅ |
-|`PHOENIX_ERROR`|**TODO**|✅ |✅ |
+|`PHOENIX_CONCURRENCY_ERROR`|**TODO**|✅ |❌ |
```
Comment (GouravSna) [inline docs/events.md]: Give example of this.

File: docs/player-functions.md [hunk 5]
```diff
@@ -60,29 +61,57 @@ Load the media with the given **assetId** OR **mediaId** OR **entryID** for OVP/
...
  ```json
- "other":"other",
+ "startPosition": 120,
   "metaData": {
     "key": "value"
   } 
  ```
-* `externalSubtitleList` *JSONArray* - 🔴TODO
-* `externalVttThumbnailUrl` - Pass the VTT thumbnail URL. These images will be shown when the user scrubs on the seekbar. To show the images logic should be handled by the FE app. You will recieve the additional information the `ImageTracks` of `TracksAvailable` event. [Check PlayerEvents](./player-event.md)
+* `externalSubtitleList` *JSONArray* - _**Optional**_  🔴TODO
```
Comment (GouravSna) [inline docs/player-functions.md]: Give example for external subtitle List

Comment (GouravSna) [reply docs/player-functions.md]: It's missing in OVP/OTT, Check it and create a ticket.
externalVttThumbnailUrl is also not there for ovp/ott

File: docs/player-functions.md [hunk 6]
```diff
@@ -60,29 +61,57 @@ Load the media with the given **assetId** OR **mediaId** OR **entryID** for OVP/
...
+* `externalVttThumbnailUrl` - _**Optional**_  Pass the VTT thumbnail URL. These images will be shown when the user scrubs on the seekbar. To show the images logic should be handled by the FE app. You will recieve the additional information the `ImageTracks` of `TracksAvailable` event. [Check PlayerEvents](./player-event.md)
```
Comment (GouravSna) [inline docs/player-functions.md:126]: fix event.md link

File: example/src/components/TrackList.tsx [hunk 7]
```diff
@@ -13,7 +13,7 @@ export default function TrackList(props) {
     labelFieldName = 'bitrate';
   } else if (trackType === 'audio') {
     incomingTrackList.map((val, idx) => val.isAdaptive === true ? val.bitrate = 'Adaptive' : val);
-    labelFieldName = 'language';
```
Comment (GouravSna) [inline example/src/components/TrackList.tsx:16]: If there is  no label, take the language

File: ios/Basic/BasicKalturaPlayerRN.swift [hunk 8]
```diff
@@ -43,20 +43,22 @@ class BasicKalturaPlayerRN: KalturaPlayerRN {
...
-                                                 base64EncodedCertificate: base64EncodedCertificate)]
-                case .widevineCenc:
-                    break
-                case .playreadyCenc:
-                    break
-                case .widevineClassic:
-                    break
-                case .unknown:
-                    break
+            if (!assetDRMData.isEmpty) {
```
Comment (GouravSna) [inline ios/Basic/BasicKalturaPlayerRN.swift]: Cross check once again.

File: src/index.tsx [hunk 9]
```diff
@@ -444,16 +446,28 @@ export class KalturaPlayerAPI {
...
-  static enableDebugLogs = (enabled: boolean) => {
-    if (enabled == null) {
+  static enableDebugLogs = (enabled: boolean, logLevel: LOG_LEVEL = LOG_LEVEL.DEBUG) => {
+    if (enabled == null || logLevel == null) {
       return;
     }
+
     debugLogs = enabled;
+
+    if (debugLogs && logLevel != LOG_LEVEL.OFF) {
```
Comment (GouravSna) [inline src/index.tsx]: Cross validate this logic again.

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt (hunk 1 above)
Comment (GouravSna) [reply android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt:547]: Done

File: docs/player-functions.md (hunk 6 above)
Comment (GouravSna) [reply docs/player-functions.md:126]: Done

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt (hunk 3 above)
Comment (GouravSna) [reply android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt]: Fixed this issue. If StartPosition is given then only consider it else don't pass it to the player.

File: ios/Basic/BasicKalturaPlayerRN.swift (hunk 8 above)
Comment (GouravSna) [reply ios/Basic/BasicKalturaPlayerRN.swift]: It's fine.

File: src/index.tsx (hunk 9 above)
Comment (GouravSna) [reply src/index.tsx]: Fixed it.

**Review by giladna — APPROVED** [review]

File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt (hunk 2 above)
Comment (GouravSna) [reply android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerRN.kt:863]: FEC-12714

File: docs/player-functions.md (hunk 5 above)
Comment (GouravSna) [reply docs/player-functions.md]: FEC-12715

File: example/src/components/TrackList.tsx (hunk 7 above)
Comment (GouravSna) [reply example/src/components/TrackList.tsx:16]: Fixed

**Review by chausov — APPROVED** [review]

**Pull request merged**